internal/expr/prepare.go

```diff
@@ -77,14 +80,63 @@ func prepareInput(ti typeNameToInfo, p *inputPart) (tm typeMember, err error) {
 			return nil, fmt.Errorf(`type %q not passed as a parameter (have "%s")`, p.sourceType.prefix, strings.Join(ts, `", "`))
 		}
 	}
-
+	if _, ok = info.(*sliceInfo); ok {
+		return nil, fmt.Errorf(`cannot use slice type %q outside of IN clause`, info.typ().Name())
+	}
+	if p.sourceType.name == "*" {
+		return nil, fmt.Errorf(`star type not allowed in standalone input`)
+	}
 	tm, err = info.typeMember(p.sourceType.name)
 	if err != nil {
 		return nil, err
 	}
 	return tm, nil
 }
 
+// prepareIn check input expressions following IN statements correspond to known
+// types and generates information about the query arguments.
+func prepareIn(ti typeNameToInfo, p *inPart) (typeMembers []typeMember, err error) {
+	defer func() {
+		if err != nil {
+			err = fmt.Errorf("input expression: %s: %s", err, p.raw)
+		}
+	}()
+	for _, t := range p.types {
+		info, ok := ti[t.prefix]
+		if !ok {
+			ts := getKeys(ti)
+			if len(ts) == 0 {
+				return nil, fmt.Errorf(`type %q not passed as a parameter`, t.prefix)
+			} else {
+				// "%s" is used instead of %q to correctly print double quotes within the joined string.
+				return nil, fmt.Errorf(`type %q not passed as a parameter (have "%s")`, t.prefix, strings.Join(ts, `", "`))
+			}
+		}
+		var tm typeMember
+		if t.name == "*" {
+			switch info := info.(type) {
+			case *structInfo, *mapInfo:
+				return nil, fmt.Errorf(`cannot use %s %q with asterisk in input expression`, info.typ().Kind(), t.prefix)
+			case *sliceInfo:
```

Comment:
So we index the slice with `$Slice.*`?  What about specific positions?  `$Slice.3`, `$Slice.[3]`?  Maybe not for now, but do we have a clear path for when that's requested?

Reply:
See update comment below